p4-fusion/main.cc

```diff
@@ -152,10 +152,66 @@ int Main(int argc, char** argv)
 		}
 		changes = std::move(changesRes.GetChanges());
 	}
+
+	// Load labels
+	PRINT("Requesting labels from the Perforce server")
+	LabelsResult labelsRes = p4.Labels();
+	if (labelsRes.HasError())
+	{
+		ERR("Failed to retrieve labels for mapping: " << labelsRes.PrintError())
+		return 1;
+	}
+	const std::list<std::string>& labels = labelsRes.GetLabels();
+	SUCCESS("Received " << labels.size() << " labels from the Perforce server")
+
+	std::unordered_map<std::string, std::unordered_map<std::string, LabelResult>*> revToLabel;
+
+	for (auto& label : labels)
+	{
+		LabelResult labelRes = p4.Label(label);
+		if (labelRes.HasError())
+		{
+			ERR("Failed to retrieve label details: " << labelRes.PrintError());
+			continue;
+		}
+		if (!labelRes.revision.starts_with("@"))
+		{
+			continue;
+		}
+		labelRes.revision.erase(labelRes.revision.begin());
+		if (labelRes.views.empty())
+		{
+			if (!revToLabel.contains(labelRes.revision))
+			{
+				auto* newMap = new std::unordered_map<std::string, LabelResult>;
+				revToLabel.insert({ labelRes.revision, newMap });
+			}
+			auto res = revToLabel.at(labelRes.revision);
+			res->insert({ convertLabelToTag(labelRes.label), labelRes });
+		}
+		else
+		{
+			for (auto& view : labelRes.views)
+			{
+				if (depotPath.starts_with(trimSuffix(view, "...")))
+				{
+					if (!revToLabel.contains(labelRes.revision))
+					{
+						auto* newMap = new std::unordered_map<std::string, LabelResult>;
+						revToLabel.insert({ labelRes.revision, newMap });
+					}
+					auto res = revToLabel.at(labelRes.revision);
+					res->insert({ convertLabelToTag(labelRes.label), labelRes });
+				}
+			}
+		}
```

Comment:
I wonder if this chunk could be somehow encapsulated better into the label_conversions file so that we don't export so many until functions in there and make main even longer - but not a blocker